gramps/gen/filters/rules/test/person_rules_test.py

```diff
@@ -1089,6 +1090,37 @@ def test_isdefaultperson(self):
             ),
         )
 
+    @patch("gramps.gen.filters.rules.person.IsDefaultPerson.apply_to_one")
+    def test_isdefaultperson_optimized(self, mock):
+        """
+        Test IsDefaultPerson rule.
+        """
+        mock.return_value = True
+        rule = IsDefaultPerson([])
+        self.assertEqual(
+            self.filter_with_rule(rule),
+            set(
+                [
+                    "GNUJQCL9MD64AM56OH",
+                ]
+            ),
+        )
+        # This used the optimizer, so it didn't loop through DB
+        self.assertEqual(mock.call_count, 1)
+
+    @patch("gramps.gen.filters.rules.person.IsFemale.apply_to_one")
+    def test_isfemale_not_optimized(self, mock):
+        """
+        Test IsFemale rule.
+        """
+        # Make everyone a female for this test:
+        mock.return_value = True
+        rule = IsFemale([])
+        # too many to list out to test explicitly
+        self.assertEqual(len(self.filter_with_rule(rule)), 2128)
+        # This did not use the optimizer, so it did loop through DB
+        self.assertEqual(mock.call_count, 2128)
```

Comment:
```suggestion
        number_of_people = 2128
        self.assertEqual(len(self.filter_with_rule(rule)), number_of_people)
        # This did not use the optimizer, so it did loop through DB
        self.assertEqual(mock.call_count, number_of_people)
```
Is this a bit clearer? (I am assuming it's one call per person)
It is worth computing the 2128 figure from the number of people in the test set?

Reply:
Yeah, mock was too complicated to do it properly, so I refactor the call_counter into a context manager. I think it could be generally useful, and the optimizer could use more tests. Thanks!

Reply:
Nice ❤️ 
I agree that this feels more generally useful but conscious that this PR really needs to get merged before it takes on a life of its own

Reply:
I plan to review, merge and probably release a beta over the next couple of days.